src/util-decode-mime.c

```diff
@@ -1026,88 +1024,101 @@ static int FindUrlStrings(const uint8_t *line, uint32_t len,
 {
     int ret = MIME_DEC_OK;
     MimeDecEntity *entity = (MimeDecEntity *) state->stack->top->data;
-    uint8_t *fptr, *remptr, *tok = NULL, *tempUrl;
-    uint32_t tokLen = 0, i, tempUrlLen;
-    uint8_t urlStrLen = 0, flags = 0;
+    MimeDecConfig *mdcfg = MimeDecGetConfig();
+    uint8_t *fptr, *remptr, *tok = NULL, *tempUrl, *urlHost;
+    uint32_t tokLen = 0, i, tempUrlLen, urlHostLen;
+    uint8_t schemeStrLen = 0, flags = 0;
+    ConfNode *scheme = NULL;
+    char *schemeStr = NULL;
+
+    if (mdcfg != NULL && mdcfg->extract_urls_schemes == NULL) {
+        SCLogDebug("Error: MIME config extract_urls_schemes was NULL.");
+        return MIME_DEC_ERR_DATA;
+    }
 
-    remptr = (uint8_t *)line;
-    do {
-        SCLogDebug("Looking for URL String starting with: %s", URL_STR);
-
-        /* Check for token definition */
-        fptr = FindBuffer(remptr, len - (remptr - line), (uint8_t *)URL_STR, strlen(URL_STR));
-        if (fptr != NULL) {
-
-            urlStrLen = strlen(URL_STR);
-            fptr += urlStrLen; /* Start at end of start string */
-            tok = GetToken(fptr, len - (fptr - line), " \"\'<>]\t", &remptr,
-                    &tokLen);
-            if (tok == fptr) {
-                SCLogDebug("Found url string");
-
-                /* First copy to temp URL string */
-                tempUrl = SCMalloc(urlStrLen + tokLen);
-                if (unlikely(tempUrl == NULL)) {
-                    SCLogError(SC_ERR_MEM_ALLOC, "Memory allocation failed");
-                    return MIME_DEC_ERR_MEM;
-                }
+    TAILQ_FOREACH (scheme, &mdcfg->extract_urls_schemes->head, next) {
```

Comment:
have you considered using mpm (multi pattern matching) here instead? That way we'd avoid searching a buffer 4 times if we have 4 patterns defined.

Reply:
Decided the simple loop here for simplicity but mpm is a good idea :) Do you have any examples from the suricata source code of how to implement mpm in C?

Reply:
The parser already uses Mpm to match on response codes. Check the src/app-layer-smtp.c file for
```c
static int SMTPProcessReply(SMTPState *state, Flow *f,
                            AppLayerParserState *pstate,
                            SMTPThreadCtx *td)
{
   ...
    PmqReset(td->pmq);
    int mpm_cnt = mpm_table[SMTP_MPM].Search(smtp_mpm_ctx, td->smtp_mpm_thread_ctx,
                                             td->pmq, state->current_line,
                                             3);
    if (mpm_cnt == 0) {
        /* set decoder event - reply code invalid */
        SMTPSetEvent(state, SMTP_DECODER_EVENT_INVALID_REPLY);
        SCLogDebug("invalid reply code %02x %02x %02x",
                state->current_line[0], state->current_line[1], state->current_line[2]);
        SCReturnInt(-1);
    }
    enum SMTPCode reply_code = smtp_reply_map[td->pmq->rule_id_array[0]].enum_value;
```

The setup logic is in `SMTPSetMpmState`.

Reply:
I took at look at how Suricata does MPM and correct me if I'm wrong but it seems that you can only see which pattern matched from the prefilter object pmq and the count of matches in the buffer, but you can't get indexes within the buffer where the pattern matched. This existing URL extraction code for SMTP needs to find the starting index of the match to properly extract the URL so I'm confused about how to use Suricata's MPM implementation here.